theano/scan_module/scan_opt.py

```diff
@@ -744,6 +744,18 @@ def process_node(self, fgraph, node):
                             self.push_out_inner_vars(fgraph, inner_dot_inputs,
                                                      node, args)
 
+                        # If truncate_gradient is used, then some sequences
+                        # don't have the same length as args.n_steps. Thus,
+                        # it causes a shape mismatch in the dot product
+                        # pushed outside of scan. Simply cutting the sequences
+                        # breaks scanOp_save_mem. Since saving memory is the
+                        # main reason for using truncate_gradient, we don't
+                        # apply the push when truncate_gradient is used.
+                        if outer_dot_inputs[0].shape[0] != args.n_steps:
```

Comment:
This won't work .shape[0] will always return new Theano variable. So you end up always disabling this optimization.

What about just doing something like:
```
if node.op.truncate_gradient != -1:
    return
```

Reply:
@nouiz: The thing is that, if I'm not mistaken, when you set `truncate_gradient`, you set it for both scans (the scan of the forward computation, and the scan for the gradient computation). I was thinking that during the forward computation, we won't have the shape problem, so we can apply this optimization. It is only during the gradient computation that we might have some issues.

So it would be nice to detect if the scan is computing the forward computation (in which case we apply the opt) or in the gradient computation (in which case we don't apply it). Is there a way to know which one is which?

Reply:
Can check your if there is mitmot? If so, it is sure this is a grad. Otherwise, it isn't sure what it is. If that is enough, then that would be great.